### ISSUE TITLE
SUBSCRIBE_DONE logic is gruesome

### ISSUE THREAD

**Issue by martinduke** [moq-wg/moq-transport#465]
How does a relay know when to send SUBSCRIBE_DONE?

The purpose of relay is to aggregate subscribes, so if overlapping subscribes come in with different endpoints I can't rely on a SUBSCRIBE_DONE from upstream.

If a SUBSCRIBE is entirely for future objects, the easiest thing to do is to send it when I've sent the last object ID, or a higher ID is published. One can add a timer per subscribe to allow for reordering, or just accept that reordered objects arriving after the last won't be delivered. I can't simply keep state for each object in the subscribe because I have no idea where there are object or group gaps. (maybe this is the use case for explicitly communicating these gaps after all?)

If a SUBSCRIBE is entirely for past objects, either I am serving them from cache (in which case I can provide an API to the cache to call when it is fully served), or I have to send a SUBSCRIBE upstream. If the SUBSCRIBE is unique, then I'll get a SUBSCRIBE_DONE from upstream. If it's aggregated somehow, once again I have to either watch for the last sequence number and run a timer, or keep score of all the objects I expect to see. IIUC, as it stands there is no guarantee that a subscribe for past objects will give me the complete set of objects.

If a SUBSCRIBE is for both past and future objects, then the union of both algorithms must both return true in order for the relay to send SUBSCRIBE_DONE. So the sender has to keep track of two separate object ranges independently. Because the SUBSCRIBE may arrive just as the publisher is incrementing the group ID, I need to track the max object ID in each group so I can tell when the past portion of the subscription is done.

At the origin, this is somewhat simpler in that published objects can be assumed to arrive in order and without gaps, and past objects can arrive in their own order and without gaps.

**Comment by kixelated** [comment]
+1 I can't figure out when a publisher is supposed to send this message. Is it supposed to wait until all data streams have been acknowledged or reset? That's not possible with WebTransport (JS) so the best you can do is a timer.

For transfork, I put the subscriber in control of terminating a subscription when all data (within the requested range) has been received. It does require drop notifications though (or reliable reset).

**Comment by afrind** [comment]
Individual Comment:

> maybe this is the use case for explicitly communicating these gaps after all?

> as it stands there is no guarantee that a subscribe for past objects will give me the complete set of objects.

What if we added these requirements/guarantees?  A publisher is required to send every sequential group/object id in the range with no gaps (either the object or a placeholder object with the status explaining why no group/object is there), or terminate the subscription with error.  End-of-group and end-of-track markers are also required.  

It would make things much easier to reason about - I kind of like the idea.  For sure it adds some overhead - is that a dealbreaker?  It's a handful of bytes to skip and object or group or end of a group.

I'm not sure even those guarantees would solve the SUBSCRIBE_DONE problem though.  Reordering of stream-per-object streams, datagrams in general, and stream resets (undefined) still seem problematic. Maybe timers are inevitable.

**Comment by vasilvv** [comment]
I wonder if we need to separate two kinds of SUBSCRIBE_DONE.

One is "EOF" generated by the original publisher.  Those can be forwarded as-is, and don't necessarily indicate that there won't be further objects due to reordering.  SUBSCRIBE_DONE isn't really needed for finite range subscribes, since the subscriber knows where that one ends.

Another is "error, and I closed the subscribe".  That one can be generated both by the relay and the original publisher.

**Comment by martinduke** [comment]
> Individual Comment:
> 
> > maybe this is the use case for explicitly communicating these gaps after all?
> 
> > as it stands there is no guarantee that a subscribe for past objects will give me the complete set of objects.
> 
> What if we added these requirements/guarantees? A publisher is required to send every sequential group/object id in the range with no gaps (either the object or a placeholder object with the status explaining why no group/object is there), or terminate the subscription with error. End-of-group and end-of-track markers are also required.
> 
> It would make things much easier to reason about - I kind of like the idea. For sure it adds some overhead - is that a dealbreaker? It's a handful of bytes to skip and object or group or end of a group.
> 
> I'm not sure even those guarantees would solve the SUBSCRIBE_DONE problem though. Reordering of stream-per-object streams, datagrams in general, and stream resets (undefined) still seem problematic. Maybe timers are inevitable.

Yes, as it stands, the spec provides essentially no expectations for what actually arrives for a subscribe. The sender might choose to omit objects in the middle of a group even though it has those objects. There are no requirements for the order in which things are sent.

And of course, there are the different stream mappings which have their own implications; in some cases FIN or RST of a stream might provide a hint, though I know Ian is resistant to exposing these explicit QUIC signals to MoQT.

It makes coding the receiver quite challenging.

**Comment by TimEvens** [comment]
Both `SUBSCRIBE_DONE` and `ANNOUNCE_CANCEL` handling appears to be unclear and could lead to interop issues with different implementations. For example, I handled this a little differently for now due to this unclarity and how to deal with the churn it causes.  https://github.com/Quicr/libquicr/blob/main/docs/moq-implementation.md#unannounce-flow goes over this a little bit.

#### Regarding cache fulfillment
I agree with @vasilvv  that SUBSCRIBE_DONE doesn't make much sense with fetch fulfillment as it could get problematic with the various state scenarios. Instead, I think fulfillment should use per-track and upon end of track/stream the data transfer is over... regardless if you got the last group/object or not. Reason why you wouldn't get last object could be caused by a stream that is closed due to RESET for whatever reason.  Using stream per track simplifies cache fulfillment in many ways, including the issues that arise with datagrams and mixed modes of live vs cache.  If the receiver believes they didn't get all they wanted, then they should request it again... considering it should have been delivered originally, and if not it's likely due to an error or that cache didn't have all the data or was purged while it was fulfilling it. 

#### Regarding Live Data Flows and  subscription/publication states
My interpretation of the current draft-04 is that `SUBSCRIBE_DONE` purpose is for the original publisher to indicate to close a track/data flow that was subscribed by the relay/server.  The original publisher may issue SUBSCRIBE_DONE for the following:

* File/flow/data is done, such as EOF, no more data to send
* Data flow has stopped, such as camera was disconnected or otherwise user-control disabled

In the above, the EOF makes most sense for SUBSCRIBE_DONE, but it doesn't really make sense for the latter.  If a camera is disabled, what is the process to start that stream/flow again when the camera is again enabled? For example, the user is playing with a switch and it's going on/off repeatedly.... Do we really expect every subscriber to re-subscribe repeatedly for each switch? That causes unnecessary churn, which would likely result in noticeable prolonged gaps of media by each subscribing user. If so, MOQT needs to further clarify that `SUBSCRIBE_DONE = subscription END` which results in streams closing everywhere upon receiving `SUBSCRIBE_DONE`. Considering there is no pause/resume, SUBSCRIBE_DONE is the only way to indicate this, but it does mean we have a lot of churn caused by it.

Below is what I think the flow is supposed to be per the current draft-04. 

```mermaid
flowchart TD
   SD[/SUBSCRIBE_DONE/]-->RHS{"Relay Have\nSubscribers?"}
   RHS-- No --> rss["(1) Remove Subscription State"]
   rss --> End
   RHS-- Yes --> l_1(("(2) for Each\nSubscriber"))
   l_1 --> send_sd[SUBSCRIBE_DONE]
   send_sd--> cs["(3) QUIC stream? FIN close it"]
   cs -->l_1
```

1. Subscription state is removed if it exists in the relay upon `SUBSCRIBE_DONE`.  Normally in the implementation this would be done after sending `SUBSCRIBE_DONE` to each subscriber
2. Iterate over each subscriber and generate `SUBSCRIBE_DONE` using the `subscribe_id` known for that subscribe
3. If the stream exists, close it using QUIC FIN close process. 

> [!NOTE]
> Publisher received data rate does not equal the subscriber transmission rate. One or more
> subscribers may fall behind because they have slower TX rates or congestion, including priority backlog.
> This can result in subscribes having a pending TX queue that grows over time. 

Items **(2)** and **(3)** above are in question here. Should we wait to `send SUBSCRIBE_DONE` and stream close via FIN upon transmission of last/max object received by publisher or not???? For live tracks I suggest **NO**.  Just close it, it's over and should end/stop immediately.  If you didn't get the full data, then you were too slow to begin with.  For different use-cases where reliable data is being transferred and is delay tolerant, then I say **MAYBE**.   It's the **maybe** case that's complicated. Should we linger subscribes/tracks that haven't finished emptying their transmit queue? 

One way to handle this, which I see pretty easy for us to implement, is to flag the state of the subscribe track as pending subscribe_done. In this state the track will not receive any new objects, but it will finish/emptying its transmit queue... Upon the last object being sent from its queue, FIN followed by `SUBSCRIBE_DONE` would be issued. Unfortunately, this process could take several seconds, even minutes under some very bad network scenarios, so timeouts will be needed. At some point, you need to give up and say it's over and I gave you enough time to get it. Although, where's the ERROR message to indicate this timeout? 

Below is an example of the flow: 

```mermaid
flowchart TD
   SD[/SUBSCRIBE_DONE/]-->RHS{"Relay Have\nSubscribers?"}
   RHS-- No --> rss["(1) Remove Subscription State"]
   rss --> End
   RHS-- Yes --> l_1(("(2) for Each\nSubscriber"))
   l_1 --> mark_us[Mark state as subscribe done]
   mark_us -->l_1
   cs -->l_1
```

Below illustrates the flow of a new thread/process/asyncio/ type monitor that monitors subscribes and their pending done states. It's unfortunate that we have to add something to check them, but normally that's not much effort/code and it shouldn't impact scale. 

```mermaid
flowchart LR
   sl[[Check Subscribe States]] --> sp{Subscribe done pending?}
   sp-- Yes --> qm{Is TX Queue Empty?}
   qm -- Yes --> send_d[Send SUBSCRIBE_DONE]
   send_d--> cs["QUIC stream? FIN close it"]
   qm -- No --> sl
    send_d --> rs[Remove Subscribe State]
```

An `UNSUBSCRIBE` could be received during the linger/pending subscribe_done state. In this case, it should be simple to handle this. If unsubscribe is received, then immediately remove the subscribe state and send a FIN close on the stream if using QUIC stream.  The other side is likely to have sent a FIN first, so race condition handling would/should take care of this at the QUIC layer.

**Comment by suhasHere** [comment]
Subscribe done serves 2 purposes 

Let's consider a topology where original publisher --> Relay --> 10000 subscribers

1. Publisher connection abruptly ends
    If relay detects connection timeout from a publisher that was actively publishing media , it can trigger "Subscribe Done" downstream to all the subscribers with appropriate condition indicating the reason for closing the subscriptions

2. Publisher ends the session normally
   In this case the publisher ends the subscription by sending "Subscribe Done" to the relay with condition indicating normal close. This can be used by relay to trigger downstream subscribers of the same

On the case when relay has been aggregating subscribes, the logic is exactly same.  A relay can aggregate subscribes only that are not overlapping. If a relay has aggregated few subscriptions to a single subscription upstream and if the publisher is ending that subscription, then the relay will close all the subscriptions that belonged to the aggregation set.

**Comment by vasilvv** [comment]
> Publisher ends the session normally
> In this case the publisher ends the subscription by sending "Subscribe Done" to the relay with condition indicating normal close. This can be used by relay to trigger downstream subscribers of the same

This sounds redundant to 0x4 object status (end of group and track).

> A relay can aggregate subscribes only that are not overlapping.

I'm not sure what you mean by that, because if I interpret that statement literally, relay would ~never be able to aggregate anything unless all clients subscribe at exactly the same time.

**Comment by suhasHere** [comment]
> > Publisher ends the session normally
> > In this case the publisher ends the subscription by sending "Subscribe Done" to the relay with condition indicating normal close. This can be used by relay to trigger downstream subscribers of the same
> 
> This sounds redundant to 0x4 object status (end of group and track).
> 

Good point. 

> > A relay can aggregate subscribes only that are not overlapping.
> 
> I'm not sure what you mean by that, because if I interpret that statement literally, relay would ~never be able to aggregate anything unless all clients subscribe at exactly the same time.

Relay should be able to aggregate most of the times except few cases. One example i can think of 

Relay has aggregated subscriptions from multiple latest group onwards for serving live edge
Then a new subscribe arrives asking for older data , absolute range , for example 

In this case, a relay need to make a new subscription request since it doesn't have any of that data ever retrieved/cached. Yes, making it a fetch explicitly will help with some of the semantics around aggregations for subscriptions. But the way we have it today, a new subscribe needs to be sent upstream in this case and cannot use aggregation.

Am i missing anything here ?

**Comment by afrind** [comment]
Individual Comment:

Thinking about using SUBSCRIBE_UPDATE to indicate a fixed endpoint for track that is currently live.  When can a relay remove this subscriber from the forwarding table?  The current logic in moxygen ends the subscription as soon as it sees an object past the endpoint, but that completely ignores the reordering case.

**Comment by fluffy** [comment]
I don't understand why we have a subscribe done at all - what is this used for?

**Comment by suhasHere** [comment]
May be it is at a point where we not seeing a real use-case for SubscribeDone . Wondering if we should remove it altogther . Happy to write a PR if others are fine ?

**Comment by ianswett** [comment]
I believe this would be fixed by #609
